agent/tasks/downloadasset.go

```diff
@@ -72,19 +72,18 @@ func (dat DownloadAssetTask) downloadAsset(url, directory string) error {
 	// TODO: What if this already exists? Consider checking file existence first with io.IsExist?
 	output, err := dat.Fs.Create(fileName)
 	if err != nil {
-		return errors.New(fmt.Sprintf("Error while creating", fileName, "-", err))
+		return fmt.Errorf("Error while creating %s - %v", fileName, err)
 	}
 	defer output.Close()
 
 	response, err := dat.HTTPClient.Get(url)
-	defer response.Body.Close()
 	if err != nil || response.StatusCode != 200 {
-
 		// If we have a problem retrieving the testlet, we want to return immediately,
 		// instead of writing an empty file to disk
-		log.Error(fmt.Sprintf("Error while downloading '%s': %s", url, response.Status))
+		log.Errorf("Error while downloading '%s': %s", url, response.Status)
 		return err
 	}
+	defer response.Body.Close()
```

Comment:
If there is an error, `response` will be nil and `response.Body.Close()` will cause a nil dereference. https://play.golang.org/p/KHNyuBKMrD

The defer should really be after checking `err`, but before checking the status code. I have another branch that breaks out the two checks, I can pull the change into this one if you'd like.